web/src/events/kernels.ts

```diff
@@ -0,0 +1,52 @@
+import { FileFormatUtils } from '@stencila/components'
+
+type DiscoverKernelsEvent = CustomEvent<{
+  kernels: string[]
+  executableLanguages: FileFormatUtils.FileFormatMap
+}>
+
+const kernelsToFileFormatMap = (
+  kernels: string[]
+): FileFormatUtils.FileFormatMap =>
+  kernels.reduce((formats: FileFormatUtils.FileFormatMap, kernelName) => {
+    const foundFormat = FileFormatUtils.lookupFormat(kernelName)
+    return { ...formats, [foundFormat.name]: foundFormat }
+  }, {})
+
+/**
+ * Emit a CustomEvent with the name of`'stencila-discover-kernels'` detailing
+ * available executable kernels. Custom WebComponents listen to this event and
+ * update the list of executable languages in the language selector fields.
+ *
+ * We also update the global `window.stencilaWebClient.executableLanguages` field
+ * so that any components added after the event has fired can be instantiated with
+ * an up to date list of languages.
+ */
+export const onDiscoverKernels = (
```

Comment:
@nokome: would be appreciated if you could let me know if you have any comments or feedback on this API or the usage/interaction between this, the various `view/edit/read.index.js`, and our custom components.